sklearn/neighbors/nca.py

```diff
@@ -0,0 +1,596 @@
+
+    init : string or numpy array, optional (default='auto')
+        Initialization of the linear transformation. Possible options are
+        'auto', 'pca', 'lda', 'identity', 'random', and a numpy array of shape
+        (n_features_a, n_features_b).
+
+        'auto':
+            Depending on ``n_components``, the most reasonable initialization
+            will be chosen among the following ones. First, we try to use
+            'lda', as it uses labels information: if ``n_components <=
+            n_classes``, ``init='lda'``. If we can't, we then try 'pca', as it
+            projects data in meaningful directions (those of higher variance):
+            if ``n_components < min(n_features, n_samples)``, ``init = 'pca'``.
+            Otherwise, we just use 'identity'.
+
+        pca:
+            ``n_components`` many principal components of the inputs passed
+            to :meth:`fit` will be used to initialize the transformation.
+            (See :class:`~sklearn.decomposition.PCA`)
+
+        lda:
+            ``min(n_components, n_classes)`` many most discriminative
+            components of the inputs passed to :meth:`fit` will be used to
+            initialize the transformation. (If ``n_components > n_classes``,
+            the rest of the components will be zero.) (See
+            :class:`~sklearn.discriminant_analysis.LinearDiscriminantAnalysis`)
+
+        identity:
+            If ``n_components`` is strictly smaller than the
+            dimensionality of the inputs passed to :meth:`fit`, the identity
+            matrix will be truncated to the first ``n_components`` rows.
+
+        random:
+            The initial transformation will be a random array of shape
+            (n_components, n_features). Each value is sampled from the
+            standard normal distribution.
+
+        numpy array:
+            n_features_b must match the dimensionality of the inputs passed to
+            :meth:`fit` and n_features_a must be less than or equal to that.
+            If ``n_components`` is not None, n_features_a must match it.
+
+    warm_start : bool, optional, (default=False)
+        If True and :meth:`fit` has been called before, the solution of the
+        previous call to :meth:`fit` is used as the initial linear
+        transformation (``n_components`` and ``init`` will be ignored).
+
+    max_iter : int, optional (default=50)
+        Maximum number of iterations in the optimization.
+
+    tol : float, optional (default=1e-5)
+        Convergence tolerance for the optimization.
+
+    callback : callable, optional (default=None)
+        If not None, this function is called after every iteration of the
+        optimizer, taking as arguments the current solution (transformation)
+        and the number of iterations. This might be useful in case one wants
+        to examine or store the transformation found after each iteration.
+
+    store_opt_result : bool, optional (default=False)
```

Comment:
Looking at this ![discussion](https://github.com/scikit-learn/scikit-learn/pull/8602#discussion_r160139021), it seems that having this result (and an option to store it or not) would be useful if we want to inspect the gradient (jacobian) at termination for instance. But now that I am thinking about it, maybe the norm of the gradient would be a sufficient information ? Or maybe the norm of the gradient normalized by its dimension, just to have a standard measure to help debug in case of wrong convergence ? If so we could indeed keep a result closer to usual scikit-learn objects like `ConvergenceWarning`, `n_iter_` etc ?

Reply:
I think a diagnostic that summarises the quality of the fit is more practical for most users.

Reply:
I agree, and now that I'm thinking about it, users who really would want more info about the convergence could always give a `callback` that would store all info needed (to see the evolution of the gradient, etc)
So I guess returning a `ConvergenceWarning` if there is no convergence, with the error message from `scipy.optimize` as is now would be sufficient ?

Reply:
Seems reasonable

Reply:
Done, in commit https://github.com/scikit-learn/scikit-learn/pull/10058/commits/49189c69c20743426dea83ba0e406a519776478e